File: dp_memoization/binary_tree_path/binary_tree_path.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BinaryTreeMaxPathSum // LeetCode Q.124.
{
private:
    int global_max_path_sum = -1000; // Min node value stated by question.

    int dfs_path_sum(TreeNode *node)
    {
        int left_child_path_sum = 0;
        int right_child_path_sum = 0;

        if (node->left != nullptr)
        {
            left_child_path_sum += max(dfs_path_sum(node->left), 0);
        }
        if (node->right != nullptr)
        {
            right_child_path_sum += max(dfs_path_sum(node->right), 0);
        }

        int left_end_path_sum = left_child_path_sum + node->val;
        int middle_path_sum = left_end_path_sum + right_child_path_sum;
        int right_end_path_sum = node->val + right_child_path_sum;

        int max_path_sum = max(
            max(left_end_path_sum, middle_path_sum), right_end_path_sum);

        if (max_path_sum > global_max_path_sum)
        {
            global_max_path_sum = max_path_sum;
        }

        return max(max(left_end_path_sum, right_end_path_sum), 0);
    }

public:
    int find_max_path_sum(TreeNode *root)
    {
        dfs_path_sum(root);
        return global_max_path_sum;
    }
};

File: dp_memoization/binary_tree_path/binary_tree_path.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreeMaxPathSum:  # LeetCode Q.124.
    def __init__(self):
        self.max_path_sum = -1000  # Min node value stated by question.

    def find_max_path_sum(self, root: TreeNode | None):
        self._dfs_path_sum(root)
        return self.max_path_sum

    def _dfs_path_sum(self, node: TreeNode) -> int:
        left_child_path_sum, right_child_path_sum = 0, 0
        if node.left:
            left_child_path_sum += max(self._dfs_path_sum(node.left), 0)
        if node.right:
            right_child_path_sum += max(self._dfs_path_sum(node.right), 0)

        left_end_path_sum = left_child_path_sum + node.val
        middle_path_sum = left_end_path_sum + right_child_path_sum
        right_end_path_sum = node.val + right_child_path_sum

        max_path_sum = max(
            left_end_path_sum, middle_path_sum, right_end_path_sum
        )
        if max_path_sum > self.max_path_sum:
            self.max_path_sum = max_path_sum

        return max(left_end_path_sum, right_end_path_sum, 0)
